### PR TITLE
feat: allow json froms with field names that are malformed variable n…

### PR DESCRIPTION
**Pull request by Jshhhh**
…ames

*Issue #, if available:*

*Description of changes:*
- Allow JSON forms with fields with invalid form names
` { "123field Name": 123}`

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/codegen-ui-react/lib/forms/form-renderer-helper.ts [hunk 1]
```diff
@@ -194,11 +199,17 @@ export const addFormAttributes = (component: StudioComponent | StudioComponentCh
...
+            undefined,
+            getCurrentValueIdentifier(fieldConfig.mappedVariableName || componentName),
+          ),
         ),
         factory.createJsxAttribute(
           factory.createIdentifier('ref'),
-          factory.createJsxExpression(undefined, factory.createIdentifier(`${lowerCaseFirst(componentName)}Ref`)),
+          factory.createJsxExpression(
+            undefined,
+            factory.createIdentifier(`${lowerCaseFirst(fieldConfig.mappedVariableName || componentName)}Ref`),
```
Comment (cwoolum) [inline packages/codegen-ui-react/lib/forms/form-renderer-helper.ts:211]: We just removed the `lowerCaseFirst` functionality in @awinberg-aws 's PR. Can you merge his in and make sure everything still passes?

File: packages/codegen-ui/lib/utils/form-component-metadata.ts [hunk 2]
```diff
@@ -25,10 +25,19 @@ import {
   GenericDataSchema,
 } from '../types';
 
+// const nanoid = customAlphabet('1234567890', 5);
```
Comment (cwoolum) [inline packages/codegen-ui/lib/utils/form-component-metadata.ts:28]: Can we remove this?

**Review by cwoolum — CHANGES_REQUESTED** [review]
Small nit and please meet with Albert to make sure his CR doesn't conflict with yours.

File: packages/codegen-ui/lib/utils/form-component-metadata.ts (hunk 2 above)
Comment (Jshhhh) [reply packages/codegen-ui/lib/utils/form-component-metadata.ts:28]: Yup

File: packages/codegen-ui-react/lib/forms/form-renderer-helper.ts (hunk 1 above)
Comment (Jshhhh) [reply packages/codegen-ui-react/lib/forms/form-renderer-helper.ts:211]: Yeah, I was going to wait for his PR to merge.

**Review by cwoolum — APPROVED** [review]
LGTM

File: packages/codegen-ui-react/lib/forms/form-state.ts [hunk 3]
```diff
@@ -81,15 +83,23 @@ export const getDefaultValueExpression = (
...
+    (acc, [name, { dataType, componentType, mappedVariableName }]) => {
+      const stateName = name.split('.')[0];
+      if (!stateNameMap.has(mappedVariableName || stateName)) {
+        acc.push(
+          buildUseStateExpression(
+            mappedVariableName || stateName,
+            getDefaultValueExpression(name, componentType, dataType),
+          ),
+        );
+        stateNameMap.set(mappedVariableName || stateName, mappedVariableName || stateName);
```
Comment (SwaySway) [inline packages/codegen-ui-react/lib/forms/form-state.ts:97]: why the need for a map if key/value are the same?
If map is required can we use a variable to resolve this conditional?

**Review by SwaySway — APPROVED** [review]
aside comment lgtm

File: packages/codegen-ui/lib/utils/form-component-metadata.ts [hunk 4]
```diff
@@ -57,6 +65,20 @@ export const mapFormMetadata = (form: StudioForm, formDefinition: FormDefinition
...
         metadata.dataType = config.dataType;
       }
+      // create mapping for field name collisions
+      let filteredFieldName = filterFieldName(name);
+      let count = 1;
+      if (mappedFieldVariableNames.has(filteredFieldName.toLowerCase()) && !name.includes('.')) {
+        filteredFieldName += count;
+        while (mappedFieldVariableNames.has(filteredFieldName.toLowerCase())) {
+          count += 1;
+          filteredFieldName = filteredFieldName.slice(0, filteredFieldName.length - 1) + count;
```
Comment (awinberg-aws) [inline packages/codegen-ui/lib/utils/form-component-metadata.ts:75]: this could break if count starts getting into double digits. Rather than repeatedly appending and slicing off the count as you continue to check for conflicting names, line72 could be `const prospectiveNewName = filteredFieldName + count;` and reassigned on this line. Then you only append to `filteredFieldName` after it's certain that it will be unique.

**Review by awinberg-aws — CHANGES_REQUESTED** [review]

File: packages/codegen-ui/lib/utils/form-component-metadata.ts (hunk 4 above)
Comment (Jshhhh) [reply packages/codegen-ui/lib/utils/form-component-metadata.ts:75]: Oh yeah good catch

File: packages/codegen-ui-react/lib/forms/form-state.ts (hunk 3 above)
Comment (Jshhhh) [reply packages/codegen-ui-react/lib/forms/form-state.ts:97]: I think this was an accident. Meant to keep the Set instead

File: packages/codegen-ui/lib/utils/form-component-metadata.ts [hunk 5]
```diff
@@ -27,12 +27,20 @@ export const getFormFieldStateName = (formName: string) => {
   return [formName.charAt(0).toLowerCase() + formName.slice(1), 'Fields'].join('');
 };
 
+/**
+ * @returns true if string contains special characters except for "." , "_" and 0-9
+ */
+export const isValidVariableName = (input: string): boolean => /^[a-zA-Z_$][a-zA-Z_$0-9]*$/g.test(input);
```
Comment (awinberg-aws) [inline packages/codegen-ui/lib/utils/form-component-metadata.ts:33]: sorry last minute ask, can we get a set of unit tests against these regexes to validate that they behave as expected? It's also easier this way to see if there may be any edge cases that may have been missed.

**Review by awinberg-aws — CHANGES_REQUESTED** [review]

File: packages/codegen-ui-react/lib/forms/form-state.ts (hunk 3 above)
Comment (Jshhhh) [reply packages/codegen-ui-react/lib/forms/form-state.ts:97]: Refactoring to use variable for conditional

File: packages/codegen-ui/lib/utils/form-component-metadata.ts (hunk 5 above)
Comment (Jshhhh) [reply packages/codegen-ui/lib/utils/form-component-metadata.ts:33]: Yup. sure thing

**Review by SwaySway — APPROVED** [review]

**Review by awinberg-aws — APPROVED** [review]
